core/modules/widgets/select.js

```diff
@@ -138,7 +142,6 @@ SelectWidget.prototype.execute = function() {
 	this.selectTitle = this.getAttribute("tiddler",this.getVariable("currentTiddler"));
 	this.selectField = this.getAttribute("field","text");
 	this.selectIndex = this.getAttribute("index");
-	this.selectClass = this.getAttribute("class");
```

Comment:
I don't think, that this should be moved to the widget.js ... If we would do it that way for every other widget parameter, it would overload widget.js. ... These elements have to be passed to `this.assignDomNodeClasses(options)` in the options parameter as Jeremy suggested. 

This issue is valid for every other widget, where you removed the parameter handling from `.execute`

Reply:
Hi @pmario @BurningTreeC perhaps there is an issue here: by removing the line from `.execute()` it means that one can no longer see all the attributes at a glance. I think @pmario is suggesting that we keep lines like this, and pass `this.selectClass` to the new methods.

Reply:
That's right and if we use an `options` object, we are able to expand the parameters in the future, without the need to change the function signature

Reply:
Ok, the `assignDomNodeClasses` method calls `this.getAttribute` to get the class(es) ... I'm a bit confused what passing the class to the method should do in that regard.
The `assignDomNodeClasses` and `assignDomNodeStyles` methods call `this.getAttribute` because the styles and classes can change dynamically and we update them with low overhead.
Why should we want to get the `class` and `style` attributes in the `execute()` method and pass them to the two new methods if the methods themselves are already designed to handle that functionality?

Reply:
the line: 

```
var classes = this.getAttribute("class",fallbackClass).split(" ");
```

should be removed from the widget.js file. It doesn't belong there. The widget base-class should provide generic functionality, which should be available in any widget, that inherits it. ... But it should not do widget specific logic. ... 

IMO reading parameters that belong to a widget is widget-specific logic. ... 

The same is true for the "style" parameter. ... 

---------

The second reason is, that it is extremely confusing if parameters are missing in the widget code, because they are handled in the base-widget. ... One of the first things, we do if we need to debug a problem with widgets, is to examine the `.execute()` function of the widget, to see if something is wrong with the parameters. 

But now some parameters are handled in a completely different part of the code tree. That's not consistent.  

If we would follow your example here, we could completely eliminate all widget files and handle all the logic in 1 monster widget.js file ... But that's not what we want.

Reply:
Yes @pmario I understand that, but it was requested to refactor the two methods into widget.js

Reply:
The problem here is that things like storing the domNode in the property "domNode" is not currently enforced by the widget baseclass; it's just a convention. The refactoring that I've suggested requires that the convention be baked into the baseclass.

I think all of this may be a good deal easier when we move to the proposed declarative approach for widget definition. Until then, I'm not sure we can do this without an unacceptable amount of code duplication.

Reply:
@Jermolene , how does the declarative approach for the widget definition look like? The words themselves don't say anything to me

Reply:
There's some recent discussion here: https://github.com/Jermolene/TiddlyWiki5/discussions/5265#discussioncomment-203399 and some history here https://github.com/Jermolene/TiddlyWiki5/issues/1903#issuecomment-145280856

Reply:
Ok I think I now understand what the declarative approach is supposed to do